torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp

```diff
@@ -0,0 +1,124 @@
+  // standard library. We also replicate IValues. So we don't need to copy
+  // them during the main benchmark phase.
+  std::vector<std::vector<ModelInput>> thread_inputs(
+      config.num_calling_threads);
+  std::vector<std::vector<ModelInput>::iterator> input_iters(
+      config.num_calling_threads);
+  {
+    std::random_device seeder;
+    std::mt19937 engine(seeder());
+    TORCH_CHECK(!inputs_.empty(), inputs_.size());
+    std::uniform_int_distribution<int> dist(0, inputs_.size() - 1);
+
+    for (int thread_id = 0; thread_id < config.num_calling_threads;
+         ++thread_id) {
+      // Just in case we generate num_iters inputs for each of the threads
+      // This was if one thread does all the work we will be fine
+      for (int i = 0; i < config.num_iters + config.num_warmup_iters;
+           ++i) {
+        thread_inputs[thread_id].push_back(inputs_[dist(engine)]);
+      }
+      input_iters[thread_id] = thread_inputs[thread_id].begin();
+    }
+  }
+
+  std::mutex m;
+  std::condition_variable cv;
+  std::atomic<int64_t> initialized{0};
+  std::atomic<int64_t> finished{0};
+  std::atomic<bool> start{false};
+  std::atomic<int64_t> num_forwards{0};
+  std::vector<std::thread> callers;
+
+  for (auto thread_id = 0; thread_id < config.num_calling_threads;
+       ++thread_id) {
+    callers.emplace_back([&, thread_id]() {
+      // We use conditional variable as a barrier to make sure each thread
+      // performs required warmeup iterations before we start measuring
+      for (auto j = 0; j < config.num_warmup_iters; ++j) {
+        module_->forward(std::move(*input_iters[thread_id]));
+        ++input_iters[thread_id];
+      }
+      LOG(INFO) << "Initialized thread " << thread_id;
+      ++initialized;
+      cv.notify_all();
+      {
+        std::unique_lock<std::mutex> g(m);
+        cv.wait(g, [&]() -> bool { return start; });
+      }
+      LOG(INFO) << "Starting forward thread " << thread_id;
+      while (num_forwards.fetch_add(1) < config.num_iters) {
+        LOG(INFO) << "Starting forward pass on thread " << thread_id;
+        module_->forward(std::move(*input_iters[thread_id]));
+        ++input_iters[thread_id];
+        LOG(INFO) << "Completed forward pass on thread " << thread_id;
+      }
+      ++finished;
+      cv.notify_one();
+      LOG(INFO) << "Shutting down forward thread " << thread_id;
+    });
+  }
+
+  {
+    std::unique_lock<std::mutex> g(m);
+    cv.wait(g, [&]() { return initialized == config.num_worker_threads; });
+  }
+  start = true;
+
+  LOG(INFO) << "Starting threads";
+  auto start_time = std::chrono::high_resolution_clock::now();
+  cv.notify_all();
+  {
+    std::unique_lock<std::mutex> g(m);
+    cv.wait(g, [&]() { return finished == config.num_calling_threads; });
+  }
+  auto end_time = std::chrono::high_resolution_clock::now();
+  LOG(INFO) << "Finished benchmark";
+
+  ExecutionStats stats;
+  float total_time_ms = std::chrono::duration_cast<std::chrono::nanoseconds>(
```

Comment:
you could probably cast to something like `duration<std::chrono::milliseconds, float>` but I'm not sure how exactly it works